### PR TITLE
Fix more upstream incompatibilities

### LINKED ISSUE #2766: FAWE causing WorldGuard region flag (deny-spawn) to fail
### Server Implementation

Paper

### Server Version

1.20.5/6

### Describe the bug

After updating both WorldGuard and FAWE to 1.20.6 my deny-spawn flag was removed and upon trying to redo that flag and reloading WG this error is given:

[16:16:59 WARN]: [WorldGuard] Failed to marshal flag value for SetFlag{name='deny-spawn'}; value is [minecraft:blaze, minecraft:skeleton, minecraft:zoglin, minecraft:ghast, minecraft:hoglin, minecraft:vindicator, minecraft:magma_cube, minecraft:piglin_brute, minecraft:spider, minecraft:creeper, minecraft:warden, minecraft:wither, minecraft:evoker, minecraft:slime, minecraft:phantom, minecraft:witch, minecraft:wither_skeleton, minecraft:husk, minecraft:cave_spider, minecraft:enderman, minecraft:stray, minecraft:vex, minecraft:zombie, minecraft:drowned, minecraft:zombie_villager]
java.lang.NoSuchMethodError: 'java.lang.String com.sk89q.worldedit.registry.Keyed.id()'
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.flags.RegistryFlag.marshal(RegistryFlag.java:64) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.flags.RegistryFlag.marshal(RegistryFlag.java:31) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.flags.SetFlag.marshal(SetFlag.java:100) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.flags.SetFlag.marshal(SetFlag.java:35) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.flags.FlagUtil.marshal(FlagUtil.java:61) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.flags.FlagUtil.marshal(FlagUtil.java:50) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.managers.storage.file.YamlRegionFile.getFlagData(YamlRegionFile.java:292) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.managers.storage.file.YamlRegionFile.saveAll(YamlRegionFile.java:236) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.managers.RegionManager.save(RegionManager.java:114) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.managers.RegionManager.saveChanges(RegionManager.java:136) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at worldguard-bukkit-7.0.10-dist.jar/com.sk89q.worldguard.protection.managers.RegionContainerImpl$BackgroundSaver.run(RegionContainerImpl.java:245) ~[worldguard-bukkit-7.0.10-dist.jar:?]
        at java.base/java.util.TimerThread.mainLoop(Timer.java:566) ~[?:?]
        at java.base/java.util.TimerThread.run(Timer.java:516) ~[?:?]

### To Reproduce

1. Download the latest versions of WorldGuard and FAWE
2. Use the command: /rg flag -w "World" -h 3 region deny-spawn pig
3. Use command /wg reload
4. Error will be given in the console

### Expected behaviour

Enabling this region flag should make all the mobs stop spawning.

### Screenshots / Videos

_No response_

### Error log (if applicable)

https://paste.gg/p/anonymous/95f530ce64ec48599c37814bcd587b80

### Fawe Debugpaste

https://athion.net/ISPaster/paste/view/bc96e72916ef45ec92f5c3efe82a2b87

### Fawe Version

2.10.1-SNAPSHOT-792;c095c49

### Checklist

- [X] I have included a Fawe debugpaste.
- [X] I am using the newest build from https://ci.athion.net/job/FastAsyncWorldEdit/ and the issue still persists.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by PierreSchwang**
## Overview
Move even more classes to records (or simulate them)

Fixes #2766

Tested with setting biomes, the worldguard flag from the issue and the `#true` + `#false` mask. Seems to work so far?

```[tasklist]
### Submitter Checklist
- [x] Make sure you are opening from a topic branch (**/feature/fix/docs/ branch** (right side)) and not your main branch.
- [x] Ensure that the pull request title represents the desired changelog entry.
- [x] New public fields and methods are annotated with `@since TODO`.
- [x] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/.github/blob/main/CONTRIBUTING.md).
```

### CODE REVIEW

File: worldedit-core/src/main/java/com/sk89q/worldedit/history/change/BlockChange.java [hunk 1]
```diff
@@ -34,38 +34,40 @@
...
+        checkNotNull(current);
+    }
 
     /**
      * Create a new block change.
      *
      * @param position the position
      * @param previous the previous block
-     * @param current  the current block
+     * @param current the current block
```
Comment (dordsor21) [inline worldedit-core/src/main/java/com/sk89q/worldedit/history/change/BlockChange.java:58]: .

File: worldedit-core/src/main/java/com/sk89q/worldedit/history/change/BlockChange.java [hunk 2]
```diff
@@ -34,38 +34,40 @@
  * <p>This block change does not have an {@link Extent} assigned to it because
  * one will be taken from the passed {@link UndoContext}. If the context
  * does not have an extent (it is null), cryptic errors may occur.</p>
+ *
+ * @param position the position
+ * @param previous the previous block
+ * @param current the current block
```
Comment (dordsor21) [inline worldedit-core/src/main/java/com/sk89q/worldedit/history/change/BlockChange.java:40]: codestyle, need spaced whitespace

File: worldedit-core/src/main/java/com/sk89q/worldedit/internal/expression/ExpressionHelper.java [hunk 3]
```diff
@@ -49,14 +55,18 @@ public static EvaluationException evalException(ParserRuleContext ctx, String me
...
-                message
+            positionInLine,
+            message
         );
     }
 
-    public static void checkIterations(int iterations, ParserRuleContext ctx) {
-        check(iterations <= 256, ctx, "Loop exceeded 256 iterations");
+    public static void checkIterations(int iterations, int positionInLine) {
+        check(iterations <= 256, positionInLine, "Loop exceeded 256 iterations");
```
Comment (dordsor21) [inline worldedit-core/src/main/java/com/sk89q/worldedit/internal/expression/ExpressionHelper.java:69]: Is this definitely keeping the same behaviour?

File: worldedit-core/src/main/java/com/sk89q/worldedit/registry/Keyed.java [hunk 4]
```diff
@@ -19,16 +19,33 @@
...
 package com.sk89q.worldedit.registry;
 
+import com.sk89q.worldedit.internal.util.DeprecationUtil;
+import com.sk89q.worldedit.internal.util.NonAbstractForCompatibility;
+
 /**
  * Represents an objects that can be added to a registry and referenced by an id which is unique within its registry.
  */
 public interface Keyed {
-
```
Comment (dordsor21) [inline worldedit-core/src/main/java/com/sk89q/worldedit/registry/Keyed.java:26]: Why no newline here?

File: worldedit-core/src/main/java/com/sk89q/worldedit/registry/Keyed.java [hunk 5]
```diff
@@ -19,16 +19,33 @@
...
     /**
      * The id of this object in the registry. Must be unique, and lowercase. Certain registries (e.g Namespaced ones) may have additional restrictions.
-     *
      * @return an id
```
Comment (dordsor21) [inline worldedit-core/src/main/java/com/sk89q/worldedit/registry/Keyed.java:31]: Here too

File: worldedit-core/src/main/java/com/sk89q/worldedit/session/Placement.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import com.sk89q.worldedit.IncompleteRegionException;
+import com.sk89q.worldedit.extension.platform.Actor;
+import com.sk89q.worldedit.math.BlockVector3;
+import com.sk89q.worldedit.regions.RegionSelector;
+import com.sk89q.worldedit.util.formatting.text.Component;
+import com.sk89q.worldedit.util.formatting.text.TextComponent;
+import com.sk89q.worldedit.util.formatting.text.TranslatableComponent;
+
+//FAWE: not in use (yet)
+public record Placement(PlacementType placementType, BlockVector3 offset) {
```
Comment (dordsor21) [inline worldedit-core/src/main/java/com/sk89q/worldedit/session/Placement.java:31]: Is this in the 7.3.x branch?

File: worldedit-core/src/main/java/com/sk89q/worldedit/session/PlacementType.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+import com.sk89q.worldedit.IncompleteRegionException;
+import com.sk89q.worldedit.extension.platform.Actor;
+import com.sk89q.worldedit.extension.platform.Locatable;
+import com.sk89q.worldedit.math.BlockVector3;
+import com.sk89q.worldedit.regions.RegionSelector;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+//FAWE: not in use (yet)
+public enum PlacementType {
```
Comment (dordsor21) [inline worldedit-core/src/main/java/com/sk89q/worldedit/session/PlacementType.java:31]: 7.3.x?

File: worldedit-core/src/main/java/com/sk89q/worldedit/session/Placement.java (hunk 6 above)
Comment (PierreSchwang) [reply worldedit-core/src/main/java/com/sk89q/worldedit/session/Placement.java:31]: Part of 7.3.x and main

File: worldedit-core/src/main/java/com/sk89q/worldedit/internal/expression/ExpressionHelper.java (hunk 3 above)
Comment (PierreSchwang) [reply worldedit-core/src/main/java/com/sk89q/worldedit/internal/expression/ExpressionHelper.java:69]: Seems like it, change from upstream and references to the method are updated as well (called with `.positionInLine()`)

**Review by dordsor21 — APPROVED** [review]

**Pull request merged**